Fix doctrine:mapping:import to work in Symfony 4 (no bundle)

weaverryan opened the pull request:
Hi guys!

This replaces and builds off of #779, which fixes #736.

It's VERY simple: the first argument can now be a bundle (like before) or a namespace:

A) If it's a bundle, everything works 100% like before.
B) If it's *not* a bundle, we assume it's the namespace. Then, there's no magic: we require them to explicitly pass us the path where the files should be generated. No fancy guess work.

Unfortunately, this is very difficult to test. So, I've purposefully made the diff very light and consistent. I've tested this in a REAL app, by copying the first 3 examples in the `setHelp()` to test everything. It all works perfectly.

Thanks!

weaverryan commented:
@kimhemsoe I see you added the "needs test" label. I'm not sure this is feasible - and only one of the commands in this bundle has a test at *all*. What do you think?

Ocramius commented:
Should still be tested though: the fact that other classes do not have any
tests is not a precondition to keep doing that

On Sat, 24 Mar 2018, 13:25 Ryan Weaver, <notifications@github.com> wrote:

> @kimhemsoe <https://github.com/kimhemsoe> I see you added the "needs
> test" label. I'm not sure this is feasible - and only one of the commands
> in this bundle has a test at *all*. What do you think?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/doctrine/DoctrineBundle/pull/791#issuecomment-375883560>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJakKr_8z0KA5cxEtkvN1E3oQ3EQP2Mks5thjsggaJpZM4S5Icj>
> .
>

weaverryan commented:
Fair enough :). I'm looking into it, and will hopefully have something soonish.

kimhemsoe commented:
@weaverryan You are ok with me merging https://github.com/doctrine/DoctrineBundle/pull/793 now before this? It will most likely mess your PR a bit up

weaverryan commented:
@kimhemsoe Yea, no problem - I can clean it up on my end :)

weaverryan commented on Command/ImportMappingDoctrineCommand.php:
```diff
@@ -23,17 +23,25 @@ protected function configure()
+<info>php %command.full_name% App\\\Entity annotation --path=src/Entity</info>
+
+Generate annotation mappings into the config/doctrine/ directory using App as the namespace:
+<info>php %command.full_name% App\\\Entity xml --path=config/doctrine</info>
```
The non-bundle version is totally explicit: pass the exact namespace and the exact path - no guessing or assuming (like we do with bundles). It may make the command slightly less user-friendly, but is more flexible, and it's a rare command to use anyways.

weaverryan commented on composer.json:
```diff
@@ -25,7 +25,7 @@
     "require": {
         "php": "^5.5.9|^7.0",
-        "symfony/framework-bundle": "~2.7|~3.0|~4.0",
+        "symfony/framework-bundle": "^2.7.22|~3.0|~4.0",
```
Needed so we get symfony/symfony#20442, else the commands in the test aren't registered correctly with deps=low

weaverryan commented:
@kimhemsoe I've rebased for #793 and added a new test for `ImportMappingDoctrineCommandTest`, which tests the old use-cases and the new one. 

And, tests are green 🎆 and my last commit should fix all the CS stuff with my changes for the new 4.0 :).

Majkl578 commented on Tests/Command/ImportMappingDoctrineCommandTest.php:
```diff
@@ -0,0 +1,146 @@
+    /** @var CommandTester|null */
+    private $commandTester;
+
+    public function setup()
```
protected

Majkl578 commented on Tests/Command/ImportMappingDoctrineCommandTest.php:
```diff
@@ -0,0 +1,146 @@
+        $this->commandTester = new CommandTester($command);
+    }
+
+    public function tearDown()
```
protected

Majkl578 commented on Command/ImportMappingDoctrineCommand.php:
```diff
@@ -59,19 +67,31 @@ protected function configure()
-        } else {
-            $destPath .= '/Resources/config/doctrine';
-        }
+        $type = $input->getArgument('mapping-type') ? $input->getArgument('mapping-type') : 'xml';
```
This is odd. It should either use hasArgument() or be simply `?:`.

Majkl578 commented on Command/ImportMappingDoctrineCommand.php:
```diff
@@ -59,19 +67,31 @@ protected function configure()
         }
 
+        $bundles = $this->getContainer()->getParameter('kernel.bundles');
+        if (isset($bundles[$input->getArgument('name')])) {
```
Better have multiple `$input->getArgument('name')` calls extracted into a variable.

Majkl578 commented on Command/ImportMappingDoctrineCommand.php:
```diff
@@ -59,19 +67,31 @@ protected function configure()
+        $bundles = $this->getContainer()->getParameter('kernel.bundles');
+        if (isset($bundles[$input->getArgument('name')])) {
+            $bundle    = $this->getApplication()->getKernel()->getBundle($input->getArgument('name'));
+            $namespace = $bundle->getNamespace() . '\\Entity';
```
No need to escape.

Majkl578 commented on Tests/Command/ImportMappingDoctrineCommandTest.php:
```diff
@@ -0,0 +1,146 @@
+
+    /**
+     * @expectedException \InvalidArgumentException
+     * @expectedExceptionMessageRegExp /The \-\-path option is required/
```
Escaping - is superfluous.

weaverryan commented:
Comments addressed! Tests green! Woo!

jwage commented:
:+1: @weaverryan can you squash now and then I think we can merge.

kimhemsoe commented:
@weaverryan Thanks

weaverryan commented:
Thanks guys! Sorry I didn't get the squashing done :)

kimhemsoe commented:
@weaverryan My fault, i was a bit fast on the trigger.
